### PR TITLE
fix: [CI-9657]: Add vanity url support

### PR DESCRIPTION
**Pull request by smjt-h**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55051)

### CODE REVIEW

**Comment by smjt-h** [comment]
trigger uts

**Comment by smjt-h** [comment]
trigger uts

File: product/log-service/handler/store.go [hunk 1]
```diff
@@ -345,3 +360,11 @@ func GetSignedURL(link, zipPrefix string, cfg config.Config, gcsClient gcputils.
 	link = link + harnessDownload + lstring[1]
 	return link, nil
 }
+
+func replaceVanityURL(vanityURL, link string) string {
+	lstring := strings.Split(link, storage)
+
+	link = vanityURL + storage + lstring[1]
```
[the remark this answers is not in the document]
Comment (smjt-h) [reply product/log-service/handler/store.go]: done

**Review by vistaarjuneja — DISMISSED** [review]

File: product/log-service/handler/store.go [hunk 2]
```diff
@@ -234,6 +238,45 @@ func HandleZipLinkPrefix(q queue.Queue, s store.Store, c cache.Cache, cfg config
...
+								WithError(err).
+								WithField("vanityurl", vanityURL).
+								WithField("Prefix", prefix).
+								Warnln("api: cannot create cache for vanity URL")
+						}
+					}
+				}
+				if vanityURL != "" {
+					link = replaceVanityURL(vanityURL, link, prefix, r)
+					logger.FromRequest(r).WithField("Prefix", prefix).Infoln("Successfully replaced with vanity url")
```
Comment (vistaarjuneja) [inline product/log-service/handler/store.go:277]: Prefix -> prefix and make info message lower case

File: product/log-service/handler/store.go [hunk 3]
```diff
@@ -234,6 +238,45 @@ func HandleZipLinkPrefix(q queue.Queue, s store.Store, c cache.Cache, cfg config
 				WriteNotFound(w, err)
 				return
 			}
+			if cfg.Platform.VanityURLEnabled {
+				//Get vanity URL from cache if it exists else calculate
+				var vanityURL string
+				exists := c.Exists(ctx, accountID+vanity)
+				if exists {
+					logger.FromRequest(r).WithField("Prefix", prefix).Infoln("vanity url exists in cache")
```
[the remark this answers is not in the document]
Comment (smjt-h) [reply product/log-service/handler/store.go:246]: done

File: product/log-service/handler/store.go [hunk 4]
```diff
@@ -345,3 +388,16 @@ func GetSignedURL(link, zipPrefix string, cfg config.Config, gcsClient gcputils.
 	link = link + harnessDownload + lstring[1]
 	return link, nil
 }
+
+func replaceVanityURL(vanityURL, link, prefix string, r *http.Request) string {
+	lstring := strings.Split(link, storage)
```
[the remark this answers is not in the document]
Comment (smjt-h) [reply product/log-service/handler/store.go:393]: done
using strings.SplitN

File: product/log-service/handler/store.go [hunk 5]
```diff
@@ -234,6 +238,45 @@ func HandleZipLinkPrefix(q queue.Queue, s store.Store, c cache.Cache, cfg config
...
+					if err != nil {
+						logger.FromRequest(r).
+							WithError(err).
+							WithField("Prefix", prefix).
+							Errorln("api: cannot fetch vanityURL from cache")
+					} else {
+						vanityURL = string(vanityURLBytes)
+					}
+				} else {
+					logger.FromRequest(r).WithField("Prefix", prefix).Infoln("Vanity URL does not exists in cache fetching from Platform")
```
[the remark this answers is not in the document]
Comment (smjt-h) [reply product/log-service/handler/store.go:257]: done

File: product/log-service/handler/store.go [hunk 6]
```diff
@@ -234,6 +238,45 @@ func HandleZipLinkPrefix(q queue.Queue, s store.Store, c cache.Cache, cfg config
...
+				//Get vanity URL from cache if it exists else calculate
+				var vanityURL string
+				exists := c.Exists(ctx, accountID+vanity)
+				if exists {
+					logger.FromRequest(r).WithField("Prefix", prefix).Infoln("vanity url exists in cache")
+					vanityURLBytes, err := c.Get(ctx, accountID+vanity)
+					if err != nil {
+						logger.FromRequest(r).
+							WithError(err).
+							WithField("Prefix", prefix).
```
[the remark this answers is not in the document]
Comment (smjt-h) [reply product/log-service/handler/store.go:251]: done

File: product/log-service/handler/store.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (smjt-h) [reply product/log-service/handler/store.go:257]: done

File: product/log-service/handler/store.go [hunk 7]
```diff
@@ -234,6 +238,45 @@ func HandleZipLinkPrefix(q queue.Queue, s store.Store, c cache.Cache, cfg config
...
+						logger.FromRequest(r).
+							WithError(err).
+							WithField(usePrefixParam, prefix).
+							Warnln("api: cannot fetch the vanity url")
+					} else {
+						err = c.Create(ctx, accountID+vanity, vanityURL, cfg.Platform.VanityURLTTL)
+						if err != nil {
+							logger.FromRequest(r).
+								WithError(err).
+								WithField("vanityurl", vanityURL).
```
[the remark this answers is not in the document]
Comment (smjt-h) [reply product/log-service/handler/store.go:269]: done

File: product/log-service/handler/store.go (hunk 2 above)
Comment (smjt-h) [reply product/log-service/handler/store.go:277]: done

File: product/log-service/handler/store.go [hunk 8]
```diff
@@ -234,6 +238,45 @@ func HandleZipLinkPrefix(q queue.Queue, s store.Store, c cache.Cache, cfg config
 				WriteNotFound(w, err)
 				return
 			}
+			if cfg.Platform.VanityURLEnabled {
+				//Get vanity URL from cache if it exists else calculate
+				var vanityURL string
+				exists := c.Exists(ctx, accountID+vanity)
```
[the remark this answers is not in the document]
Comment (smjt-h) [reply product/log-service/handler/store.go]: done moved it to separate func

File: product/log-service/handler/store.go [hunk 9]
```diff
@@ -345,3 +383,30 @@ func GetSignedURL(link, zipPrefix string, cfg config.Config, gcsClient gcputils.
 	link = link + harnessDownload + lstring[1]
 	return link, nil
 }
+
+func replaceVanityURL(vanityURL, link, prefix string, r *http.Request) string {
```
Comment (vistaarjuneja) [inline product/log-service/handler/store.go:387]: Nit: don't pass in the http request into functions like these. It makes it much harder to test in the future. These functions should just take simple inputs and produce simple outputs.

Comment (vistaarjuneja) [reply product/log-service/handler/store.go:387]: Logging can be done at the top level and functions can return accordingly

**Review by vistaarjuneja — DISMISSED** [review]

File: product/log-service/handler/store.go (hunk 9 above)
Comment (smjt-h) [reply product/log-service/handler/store.go]: done

**Review by vistaarjuneja — APPROVED** [review]

**Pull request merged**